flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsProjectionPushDown.java

```diff
@@ -0,0 +1,72 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.connector.source.abilities;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.connector.source.ScanTableSource;
+
+/**
+ * Enables to push down a (possibly nested) projection into a {@link ScanTableSource}.
+ *
+ * <p>Given the following SQL:
+ * <pre>{@code
+ *   CREATE TABLE t (i INT, r ROW < d DOUBLE, b BOOLEAN>, s STRING);
+ *   SELECT s, r.d FROM t;
+ * }</pre>
+ *
+ * <p>In the above example, {@code r.d} and {@code s} are required fields. Other fields can be skipped
+ * in a projection.
+ *
+ * <p>By default, if this interface is not implemented, a projection is applied in a subsequent operation
+ * after the source.
+ *
+ * <p>For efficiency, a source can push a projection further down in order to be close to the actual
+ * data generation. A projection is only selecting fields that are used by a query. It does neither
+ * reorder fields nor contain any computation. A projection can either be performed on the fields of
+ * the top-level row only or consider nested fields as well (see {@link #supportsNestedProjection()}).
+ *
+ * <p>Note: If a source implements {@link SupportsComputedColumnPushDown}, the projection must be applied
+ * to the physically produced data only. {@link SupportsComputedColumnPushDown} assumes an already applied
+ * projection and includes a projection of computed columns.
+ */
+@PublicEvolving
+public interface SupportsProjectionPushDown {
+
+	/**
+	 * Returns whether this source supports nested projection.
+	 */
+	boolean supportsNestedProjection();
+
+	/**
+	 * Provides the field index paths that should be used for a projection. The indices are 0-based and
+	 * support fields within (possibly nested) structures if this is enabled via {@link #supportsNestedProjection()}.
+	 *
+	 * <p>In the example mentioned in {@link SupportsProjectionPushDown}, this method would receive:
+	 * <ul>
+	 *     <li>{@code [[1], [2]]} for {@code r} and {@code s} if {@link #supportsNestedProjection()}
+	 *     returns false.
+	 *     <li>{@code [[1, 0], [2]]} for {@code r.d} and {@code s} if {@link #supportsNestedProjection()}
+	 *     returns true.
+	 * </ul>
+	 *
+	 * @param projectedFields field index paths of all fields that must be present in the physically
+	 *                        produced data
+	 */
+	void applyProjection(int[][] projectedFields);
```

Comment:
The limitation of this way is that we can't support deep nested fields projection. For example

```sql
CREATE TABLE t1 (r ROW< d DOUBLE, f ROW<q1 STRING, q2 BOOLEAN> >);

SELECT r.f.q2 FROM t1;
```

We can't prune `r.f.q1` using current API.

Reply:
With this approach we can prune deeply nested fields and even reorder those fields. The argument is simply an array of paths for selection. If we prune `r.f.q1`, its path will not be in the array. For your example the array will look like:

`[[0, 1, 1]]` which is equivalent to `[["r", "f", "q2"]]`. The latter can be simply computed out of a table schema or data type.

Reply:
Thanks for the explanation. I'm fine with this change now.